### PR TITLE
Manure excretion

### PR DESCRIPTION
**Pull request by militsasotirova**
This is the code for manure excretion calculations- I've attached the pseudocode and some more information below. For testing, I created a new json file with a partially updated feed library.

If you would like to test the code, do the following (I didn't want to push extra testing code since I was just checking if the operations were correct, and this code hasn't yet been connected with the rest of the simulation):
**In animal.py:**
- Make the import statement read: from RUFAS.routines.animal import ration, Herd, manure_excretion
- Comment out the lines where the ration requirements are calculated, the call to ration.optimize() and the setting of the ‘infeasible’ variable and add the following right after:
            _formulated_ration = {    
                'Corn_grain': 0,
                'Legume_hay': 13.5588,
                'Cotton_seed': 6.3620,
                'Roasted_soybean': 2.3964,
                'Rye_hay': 0,
                'status': 'Optimal',
                'objective': 4.5756
            }
            infeasible = False_
- Add _manure_excretion.test_manure(feed)_ after _self.ration['MP_reduction'] = milk_production_multiplier_

**In feed.py:**
- Comment out lines 59-75 (these lines are taken care of in the pull request for the ration code, but that hasn't been merged with master yet).

Pseudocode: [Manure excretion pseudocode.pdf](https://github.com/RuminantFarmSystems/MASM/files/3277657/Manure.excretion.pseudocode.pdf)

List of inputs and outputs: [Manure inputs and outputs.pdf](https://github.com/RuminantFarmSystems/MASM/files/3277660/Manure.inputs.and.outputs.pdf)

Data for testing: [Manure simulation.pdf](https://github.com/RuminantFarmSystems/MASM/files/3277658/Manure.simulation.pdf)

### CODE REVIEW

**Review by wmaxdonovan — APPROVED** [review]
Equations for manure excretion are implemented correctly and are easy to understand thanks to comments and pseudocode. In the method test_manure you create the dictionary ration_formulation that runs the for loop that calculates nutrient content of each feed type, and you have this if statement that checks whether the key is a type of feed against a list called available_feed_names, which I believe is generated from the JSON file. How is ration_formulation going to be generated outside of this specific test, and is there going to be a way to make sure that the names of those feed types are identical to the names given in the JSON input?

**Comment by militsasotirova** [comment]
> Equations for manure excretion are implemented correctly and are easy to understand thanks to comments and pseudocode. In the method test_manure you create the dictionary ration_formulation that runs the for loop that calculates nutrient content of each feed type, and you have this if statement that checks whether the key is a type of feed against a list called available_feed_names, which I believe is generated from the JSON file. How is ration_formulation going to be generated outside of this specific test, and is there going to be a way to make sure that the names of those feed types are identical to the names given in the JSON input?

The ration formulation uses the names given in the JSON file to calculate a ration and then the results of this formulation will eventually be passed to this manure excretion code once the animal module is restructured. These names will be identical because there is only one source for them.
Thanks for your comments!

**Comment by wmaxdonovan** [comment]
> > Equations for manure excretion are implemented correctly and are easy to understand thanks to comments and pseudocode. In the method test_manure you create the dictionary ration_formulation that runs the for loop that calculates nutrient content of each feed type, and you have this if statement that checks whether the key is a type of feed against a list called available_feed_names, which I believe is generated from the JSON file. How is ration_formulation going to be generated outside of this specific test, and is there going to be a way to make sure that the names of those feed types are identical to the names given in the JSON input?
> 
> The ration formulation uses the names given in the JSON file to calculate a ration and then the results of this formulation will eventually be passed to this manure excretion code once the animal module is restructured. These names will be identical because there is only one source for them.
> Thanks for your comments!

That sounds great! Totally clear then.

File: RUFAS/routines/animal/manure_excretion.py [hunk 1]
```diff
@@ -40,31 +40,71 @@ def manure_calculations(ration_formulation, BW, DIM, mPrt):
...
+            # kg of each nutrient
+            DM_feed_amount = ration_formulation[key]
+            ADF_feed_amount = ADF_feed_content * DM_feed_amount
+            CP_feed_amount = CP_feed_content * DM_feed_amount
+            LIG_feed_amount = LIG_feed_content * DM_feed_amount
+            Ash_feed_amount = Ash_feed_content * DM_feed_amount
+            NDF_feed_amount = NDF_feed_content * DM_feed_amount
+            
+            # add to running sums
+            feed_amount = DM_feed_amount / DM_feed_content
```
Comment (KFosterReed) [inline RUFAS/routines/animal/manure_excretion.py]: Can we change the name of the feed_amount variable to as_fed_feed_amount? "As fed'" is a common term used to distinguish between feed quantities based on the DM weight or the total weight (including moisture).

Comment (militsasotirova) [reply RUFAS/routines/animal/manure_excretion.py]: Yes I just committed that change. Thanks for the input!

**Comment by KFosterReed** [comment]
This looks good to me but I couldn't get the code to run with or without the changes to test the manure excretion function. I keep getting an error about inconsistent indentation. I think this is because I've been using notepad as my editor. I will revisit it tomorrow morning. So far though, everything looks great!

**Comment by militsasotirova** [comment]
Sounds good- let me know if you have questions/need help!

**Review by KFosterReed — APPROVED** [review]

**Review by manfei-L — APPROVED** [review]
It looks good to me, thanks

**Pull request merged**